### ISSUE TITLE
Code uses exceptions for normal program flow

### ISSUE THREAD

**Issue by RalfMengwasser** [influxdata/influxdb-client-csharp#109]
The code uses exceptions even for successful code executions.

Exceptions completely screw up the execution stack and you should NEVER use exceptions instead of events and you should NEVER use exceptions for anything that is not an exception that can not be handled by the code.

A returned API call of 404 or whatever is NOT an exception. Its an information that needs to be passed on to the user.

**Comment by bednar** [comment]
Hi @RalfMengwasser,

What do you think about extends `WriteApiAsync` with `Task<IRestResponse> *WithResponse` methods?

API will be looks like:

```csharp
...

public async Task WriteRecordAsync(WritePrecision precision, string record)
{
    ...
}

public async Task<IRestResponse> WriteRecordAsyncWithResponse(WritePrecision precision, string record)
{
    ...
}

...
```

You will be able to hook to `Task<IRestResponse>`.

-- https://restsharp.dev/api/RestSharp.html#interface-irestresponse

Regards

**Comment by RalfMengwasser** [comment]
Yes thats what I just did on my fork.

And I had to remove all the Object and all the String because there is no way for this to go through static code checks.

And I am in the process of removing all those Enum == null calls because there is no way an enum can be null and the warnings break the build.

It will take a while for me to test this. Too many things changed.

**Comment by bednar** [comment]
The issue is fixed in [1.19.0](https://github.com/influxdata/influxdb-client-csharp/milestone/22) milestone. You can use latest dev version by:

```console
dotnet add package InfluxDB.Client --version 1.19.0-dev.3051
```

[`WriteApiAsync`](https://github.com/influxdata/influxdb-client-csharp/blob/6636bec115c7d751c02681571e814423f87f0e53/Client/WriteApiAsync.cs#L15) now supports return `IRestResponse` for custom handling of HTTP response:

- https://github.com/influxdata/influxdb-client-csharp/blob/6636bec115c7d751c02681571e814423f87f0e53/Client/WriteApiAsync.cs#L185
- https://github.com/influxdata/influxdb-client-csharp/blob/6636bec115c7d751c02681571e814423f87f0e53/Client/WriteApiAsync.cs#L314
- https://github.com/influxdata/influxdb-client-csharp/blob/6636bec115c7d751c02681571e814423f87f0e53/Client/WriteApiAsync.cs#L484
